lib/image_optim/cache.rb

```diff
@@ -0,0 +1,72 @@
+require 'digest/sha1'
+require 'fspath'
+require 'image_optim/cache_path'
+
+class ImageOptim
+  # Handles image cache
+  class Cache
+    def initialize(image_optim, workers_by_format)
+      return unless image_optim.cache_dir
+      @cache_dir = FSPath.new(image_optim.cache_dir)
+      @cache_worker_digests = image_optim.cache_worker_digests
+      @options_by_format = Hash[workers_by_format.map do |format, workers|
+        [format, workers.map(&:inspect).sort.join(', ')]
+      end]
+      @bins_by_format = Hash[workers_by_format.map do |format, workers|
+        [format, workers.map(&:used_bins).flatten!.map! do |sym|
+          bin = image_optim.resolve_bin!(sym)
+          "#{bin.name}[#{bin.digest}]"
+        end.sort!.uniq.join(', ')]
+      end]
+    end
+
+    def fetch(original)
+      return yield unless @cache_dir
+
+      digest = digest(original, original.image_format)
+      cached = @cache_dir / digest
+      return CachePath.convert(cached) if cached.file?
+      return if cached.symlink? && File.readlink(cached) == @cache_dir.to_s
```

Comment:
Not good if cache dir is moved or if it is specified differently ('cache', './cache', '/full/path/to/cache').
This made me think about what is it better to link to. I understand that we should not care about link destination, currently it will be most probably broken if cache dir is relative. It can also be a fake location like '-' to make it very small or file digest to make it descriptive. What do you think?